extensions/wikia/SpecialDiscussionsLog/SpecialDiscussionsLogController.class.php

```diff
@@ -0,0 +1,230 @@
+						->escaped(),
+				'hasNoUserLogRecords' => $hasNoUserLogRecords,
+				'noUserLogRecordsMessage' => wfMessage( 'discussionslog-no-mobile-activity-error' )
+						->params( $userName )
+						->escaped(),
+				'timestampHeader' => wfMessage( 'discussionslog-timestamp-header' )
+						->escaped(),
+				'userAgentHeader' => wfMessage( 'discussionslog-user-agent-header' )
+						->escaped(),
+				'userLogRecords' => $displayedUserLogRecords
+			]
+		);
+	}
+
+	private function aggregateLogSearches( $userId, $userName ) {
+		$logger = Wikia\Logger\WikiaLogger::instance();
+		$query = $this->getQuery( $userId );
+		$client = new GuzzleHttp\Client();
+		$records = [];
+		$dayOffset = 0;
+		while ( count( $records ) < self::PAGINATION_SIZE && $dayOffset < self::DAYS_RANGE ) {
+			$url = $this->constructKibanaUrl( $dayOffset++ );
+			try {
+				$response = $client->post(
+					$url, [
+						'body' => $query,
+					]
+				);
+			} catch ( \GuzzleHttp\Exception\RequestException $requestException ) {
+				$logger->warning(
+					sprintf( 'Request to elasticsearch failed: %s', $requestException->getMessage() )
+				);
+				continue;
+			}
+
+			if ( $response->getStatusCode() !== self::HTTP_STATUS_OK ) {
+				$logger->warning(
+					sprintf( 'Elasticsearch request error; status code %d', $response->getStatusCode() )
+				);
+				continue;
+			}
+
+			$resultObject = json_decode( $response->getBody() );
+			$hits = $resultObject->hits->hits;
+			foreach ( $hits as $hit ) {
+				$record = $hit->_source;
+
+				$userLogRecord = new UserLogRecord();
+				$userLogRecord->setApp(
+					$record->{'mobile_app.data.app_name'} . ' ' . $record->{'mobile_app.data.app_version'}
+				);
+				$userLogRecord->setIp( $record->{'mobile_app.client_ip'} );
+				$userLogRecord->setLanguage( $record->{'mobile_app.event.device_language'} );
+				$userLogRecord->setLocation(
+					$record->{'mobile_app.geo_ip.city'} . ', ' . $record->{'mobile_app.geo_ip.country_name'}
+				);
+				$userLogRecord->setTimestamp( date( DATE_RFC2822, $record->{'mobile_app.event.timestamp'} / 1000 ) );
+				$userLogRecord->setUserAgent(
+					$record->{'mobile_app.data.platform'} . ' ' . $record->{'mobile_app.data.platform_version'}
+				);
+				$userLogRecord->setUserId( $userId );
+				$userLogRecord->setUserName( $userName );
+				$records[] = $userLogRecord;
+			}
+		}
+
+		return $records;
+	}
+
+	private function getQuery( $userId ) {
+		$pagination_size = self::PAGINATION_SIZE;
+		return <<<JSON_BODY
+{
+	"query": {
+		"filtered": {
+			"query": {
+				"bool": {
+					"should": [{
+						"query_string": {
+							"query":"mobile_app.event.user_id:$userId"
+						}
+					}]
+				}
+			}
+		}
+	},
+	"size":$pagination_size,
+	"sort":[{
+		"@timestamp": {
+			"order": "desc"
+		}
+	}]
+}
+JSON_BODY;
+	}
+
+	private function checkAccess() {
+		if ( !$this->wg->User->isLoggedIn() || !$this->wg->User->isAllowed( 'forumadmin' ) ) {
+
+			$this->wg->Out->clearHTML();
```

Comment:
Updating the output inside of this method feels like a side effect to me. Since this method is called `checkAccess`, I think that's all we should have it do and move this output logic after the the check.

Also, there's this nifty [PermissionsErrors.php](https://github.com/Wikia/app/blob/dev/includes/Exception.php#L373) which could be useful. Here's an example of its usage: [Usage](https://github.com/Wikia/app/blob/dev/extensions/wikia/Email/SpecialSendEmailController.class.php#L32).